### ISSUE TITLE
v5: The sync loader for "*.config.js" returned a Promise

### ISSUE THREAD

**Issue by shellscape** [davidtheclark/cosmiconfig#134]
Enjoying the enhanced simplicity of the v5 release. I've been able to remove quite a lot of code as a result of upgrading. However, there is one sticking point that is preventing me from moving forward: There are completely valid situations in which a module may return a `Promise` while using `searchSync`.

I present exhibit A, `webpack.config.js`:

```js
module.exports = new Promise((resolve /* reject */) => {
  resolve({
    entry: 'promise',
    mode: 'development',
  });
});
```

That `Promise` should be returned to the consumer of `cosmiconfig` so that the `Promise` can be operated on by the consumer. This pattern has been supported in `webpack` for a little over a year now, and it's a blocker for the project to leverage v5. I'm not sure what the justifications were for requiring that the result from a `searchSync` call had to be only an `Object` or `null`, but that seems on the surface like a needless limitation.

**Comment by shellscape** [comment]
> If you comment out that error condition do things work fine for you?

Indeed it does. I wrote a monkeypatch to override `loadFileContent` to do just that while a fix/change is pending.

> I think I'd be open to removing the error condition, maybe just adding a warning to the readme about how you need to think about your sync/async usage when writing custom loaders.

That could be coupled with an option to enforce strict return types for `sync`. But a helpful _Note: ..._ would probably be just as effective. I don't think there's the same level of mystery around `Promise` these days that there used to be in the past, so hopefully people using them in their configs are away of what they're doing.

**Comment by shellscape** [comment]
@davidtheclark confirmed that it works as expected in our fork, tests passing in the main project leveraging the module. please do forgive the followup question, as I know it's generally frowned up, but is there any chance of seeing this chang

**Comment by davidtheclark** [comment]
@shellscape Not sure what timezone you're in ... but I should be able to get the change done within the next ... say ... 5 hours?

**Comment by shellscape** [comment]
Cool, I'll still be conscious!

**Comment by olsonpm** [comment]
Ha - I was happy to see the `instanceof` check because I'm one of those developers that makes stupid mistakes and thus code defensively to save myself debugging time.  The webpack use-case makes perfect sense though.  Thanks for bringing that up.

**Comment by chrisblossom** [comment]
I'm confused why you'd use ``searchSync`` instead of ``search`` if you are returning a ``Promise``? ``search`` should be able to accept an ``async`` or ``sync`` loader. I must be misunderstanding something here?

**Comment by olsonpm** [comment]
yeah, the problem is the ambiguity presented when the config returned **is** a promise.

**Comment by chrisblossom** [comment]
@olsonpm My point is if ``webpack`` or any other consumer is able to handle an ``async`` config, it should use ``search``. ``searchSync`` should not be ``async``, which returning a ``Promise`` config makes it ``async``, right?

@shellscape is there a reason why ``webpack`` is using ``searchSync`` instead of ``search``?

**Comment by olsonpm** [comment]
I understood it this way: why would you prevent someone from using searchSync just because the config happens to hold a promise?

**Comment by shellscape** [comment]
> I understood it this way: why would you prevent someone from using searchSync just because the config happens to hold a promise?

Exactly.

> is there a reason why webpack is using searchSync instead of search?

Yes. The config is passed downstream and evaluated at a later time. We do not want to evaluate the `Promise` at the point at which `cosmiconfig` loads it from disk. Whether or not you agree with that approach, it's difficult to make an argument for limiting users in this situation aside from "protecting users from themselves." If that is the argument and you're ultimately passionate about it, then we need an option to turn off the type checking. Limiting users as to what their configs can contain based upon how it's loaded is awfully presumptuous.

**Comment by chrisblossom** [comment]
@olsonpm I guess to me it signals the consumer is misusing ``cosmiconfig``, and that should guide them to use it as they "should". But as @shellscape said, this might be very presumptuous.

> The config is passed downstream and evaluated at a later time. We do not want to evaluate the Promise at the point at which cosmiconfig loads it from disk.

When you say evaluate, do you mean wait (``.then``) on the ``search`` ``Promise`` to return the config?


I'm not passionate about this at all, I guess just trying to understand the issue. I just can't think of a situation where you'd be able to consume an ``async`` config, but also be forced to use ``searchSync``.

**Comment by davidtheclark** [comment]
> awfully presumptuous

I wouldn't say *awfully* presumptuous — this use case is pretty weird. But I'm fine cutting the check for now. We'll see if anybody ends up *wishing there were* such a type check in place (maybe not).

**Comment by shellscape** [comment]
> When you say evaluate, do you mean wait (.then) on the search Promise to return the config?
> When you say evaluate, do you mean wait (.then) on the search Promise to return the config?

@chrisblossom not exactly. our situation is one where we're trying to consolidate and create a "single truth" for loading webpack configs. At present there are around two dozen projects out there, both in and out of the webpack/webpack-contrib org, that search for and load config files for webpack. That includes 6 within the webpack org itself. The `config-loader` module is leveraging `cosmiconfig`, and that's our single point of truth. In the example I provided above of a minimal config that exported a `Promise`, we don't want `cosmiconfig` evaluating that export, that is to say we don't want it `.then()`ing the `Promise` the config actually contains. That's up to consumers of `config-loader` to tackle, or if we have our dirty little way later on, a series of potential hooks around that (which is all internal stuff not pertinent to this issue). The crux there is that we want to load the config's content from disk synchronously at the moment, regardless of what the export is, because we want to have our way with it later on. And we also don't have a need for it to be async, either. That may sound funky given that the config _quite possibly could be_, but 9 times out of 10 it's not. It's an obscure option that, regardless of merit or the wisdom behind it, managed to find its way into the project a while back.

@davidtheclark much appreciated. I'm good with literally any combination or outcome that results in us being able to run away with the config export from a sync search without type restrictions.

**Comment by davidtheclark** [comment]
@shellscape Published in 5.0.2.

**Comment by shellscape** [comment]
Thanks very much all 🎉
